src/box/lua/config/applier/box_cfg.lua

```diff
@@ -428,7 +438,206 @@ local function apply(config)
+    -- In such case alert was not done, it's safe to ignore such case.
+    local instance_name = nil
+    if new == nil then
+        instance_name = names_state.config._configdata:peer_name_by_uuid(old[2])
+    elseif old[3] == nil then
+        instance_name = new[3]
+    end
+
+    if instance_name ~= nil then
+        box.on_commit(function()
+            names_state.config:_alert_drop(instance_name)
+            names_check_and_clean()
+        end)
+    end
+end
+
+local function missing_names_is_empty(missing_names, replicaset_name)
+    return not missing_names[replicaset_name] and
+           table.equals(missing_names.peers, {})
+end
+
+local function names_on_rw_transit()
+    local configdata = names_state.config._configdata
+    local replicaset_name = configdata._replicaset_name
+    local missing_names = configdata:missing_names()
+
+    if missing_names_is_empty(missing_names, replicaset_name) then
+        -- Somebody have done work for us, nothing to update.
+        return
+    end
+
+    box.begin()
+    -- Set replicaset_name.
+    if missing_names[replicaset_name] ~= nil then
+        box.space._schema:insert{'replicaset_name', replicaset_name}
+    end
+
+    -- Set names for all instances in the replicaset.
+    for name, uuid in pairs(missing_names.peers) do
+        local tuple = box.space._cluster.index.uuid:select(uuid)
+        box.space._cluster:update(tuple[1][1], {{'=', 3, name}})
+    end
+    box.commit()
+end
+
+local function names_rw_watcher()
+    local channel = names_state.channel_name
+    local trigger = names_state.rw_trigger_name
+
+    if box.info.ro then
+        return box.watch('box.status', function(_, status)
+            if status.is_ro == false then
+                worker:put(channel, trigger)
+            end
+        end)
+    end
+
+    -- Immediately wakeup worker, watcher is not needed.
+    worker:put(channel, trigger)
+    return nil
+end
+
+
+local function no_missing_names_alerts(config)
+    local configdata = config._configdata
+    if config._alerts[configdata._replicaset_name] ~= nil then
+        return false
+    end
+
+    local peers = configdata:peers()
+    for _, name in ipairs(peers) do
+        if config._alerts[name] ~= nil then
+            return false
+        end
+    end
+
+    return true
+end
+
+local function names_apply()
+    local config = names_state.config
+    local missing_names = config._configdata:missing_names()
+    if missing_names_is_empty(missing_names, config._configdata) then
+        -- Exit early if all names are set.
+        return
+    end
+
+    -- Even if everything is configured we try to make alerts one
+    -- more time, as new instances without names may be found.
+    names_alert_missing(config)
+    if names_state.is_configured then
+        -- All triggers are already configured, nothing to do, but wait.
+        return
+    end
+
+    box.space._schema:on_replace(names_schema_on_replace)
```

Comment:
Don't you still need the `on_schema_init` trigger to set the `on_replace`?

I don't see why the same problem we discussed yesterday (with a name not yet being committed, but already replaced) can't happen here.

Reply:
You're right. Now `on_schema_init` is used, though it's not enough to make it safe